### PR TITLE
PRL-3954:Transfer to another court

### PR DESCRIPTION
**Pull request by khurshedk**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/PRL-3899

### Change description ###
Transfer court is a new event introduced in Private law to automate the process of having the right court to make decision on a case.

A courtadmin, judge and LA can execute this event from Case Issued state.

**Pre-requisites:** Courtadmins can draw orders when a case is issued. Therefore to transfer a case, they are supposed to draw a C21 (blank order) with either of Other or no order made option.

**Event details:** User will be presented with the list of courts which are onboarded to WA for PRL. If the user finds the relevant court, they can choose and transfer the case.

This process will keep the case online. Case state will not change. All the WA tasks allocated to the court will be auto transferred to new court location, except the assigned one (assigned ones has to be reassigned manually due to WA limitations. No email notification is in process.

If the user does not find the relevant court, i.e. case is not yet onboarded to WA, then they have to provide court name and an email address. Case will be progressed offline since then. Case state will be changed to 'Proceeds in Heritage system'. All the WA tasks will be auto configured to cancelled except the assigned ones. An email will be sent out to the email address provided stating all the documents are coming to them through another email. the next email will be sent via sendgrid, which will contain all the documents from the case. In this way, even if send grid notification fails, next court will be informed that there are documents supposed to come and they can enquire about the case.

Jira epic: https://tools.hmcts.net/jira/browse/PRL-3899
Main tkt: https://tools.hmcts.net/jira/browse/PRL-3947
Figma link: https://www.figma.com/file/zgIKniC5KvXl1unoT7obi1/Transfer-to-another-court?type=design&node-id=2-23&mode=design&t=4epuXz9aFfmNYYmq-0

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: definitions/private-law/json/State.json [hunk 1]
```diff
@@ -109,5 +109,14 @@
     "Description": "Requested for deletion",
     "DisplayOrder": 12,
     "TitleDisplay": "# #${[CASE_REFERENCE]}\n ## Case Name: ${applicantCaseName}\n ## Family Man ID: ${familymanCaseNumber}"
+  },
+  {
+    "LiveFrom": "26/04/2023",
+    "CaseTypeID": "PRLAPPS",
+    "ID": "PROCEEDS_IN_HERITAGE_SYSTEM",
+    "Name": "Proceed in heritage system",
+    "Description": "Proceed in heritage system",
```
Comment (AlokDatta) [inline definitions/private-law/json/State.json:118]: Can we change this description to say 'Proceeding in offline mode in familyman system'

File: definitions/private-law/json/CaseEvent/CaseEvent.json [hunk 2]
```diff
@@ -997,7 +997,7 @@
     "Description": "Transfer to another court",
     "DisplayOrder": 107,
     "PreConditionState(s)": "CASE_ISSUED;JUDICIAL_REVIEW",
-    "PostConditionState": "*",
+    "PostConditionState": "PROCEEDS_IN_HERITAGE_SYSTEM(cantFindCourtCheck!=\"null\"):1;*",
```
Comment (AlokDatta) [inline definitions/private-law/json/CaseEvent/CaseEvent.json:1000]: can we do this via backend call please with about to submit callback?

File: definitions/private-law/json/AuthorisationCaseState.json [hunk 3]
```diff
@@ -343,5 +343,41 @@
...
     ]
+  },
+  {
+    "CaseTypeID": "PRLAPPS",
+    "CaseStateID": "PROCEEDS_IN_HERITAGE_SYSTEM",
+    "AccessControl": [
+      {
+        "UserRoles": [
+          "caseworker-privatelaw-solicitor",
+          "citizen",
```
Comment (AlokDatta) [inline definitions/private-law/json/AuthorisationCaseState.json:354]: Only CRU for systemupdate, rest of the world only R

File: definitions/private-law/json/AuthorisationCaseState.json [hunk 4]
```diff
@@ -343,5 +343,41 @@
...
+          "caseworker-privatelaw-courtadmin",
+          "caseworker-privatelaw-la",
+          "caseworker-privatelaw-judge",
+          "caseworker-caa",
+          "caseworker-approver",
+          "ctsc-team-leader",
+          "caseworker-privatelaw-cafcass",
+          "caseworker-privatelaw-externaluser-viewonly"
+        ],
+        "CRUD": "CRUD"
```
Comment (AlokDatta) [inline definitions/private-law/json/AuthorisationCaseState.json:370]: only R

File: definitions/private-law/json/State.json [hunk 5]
```diff
@@ -109,5 +109,14 @@
     "Description": "Requested for deletion",
     "DisplayOrder": 12,
     "TitleDisplay": "# #${[CASE_REFERENCE]}\n ## Case Name: ${applicantCaseName}\n ## Family Man ID: ${familymanCaseNumber}"
+  },
+  {
+    "LiveFrom": "26/04/2023",
+    "CaseTypeID": "PRLAPPS",
+    "ID": "PROCEEDS_IN_HERITAGE_SYSTEM",
+    "Name": "Proceed in heritage system",
```
Comment (AlokDatta) [inline definitions/private-law/json/State.json:117]: Proceeds

File: definitions/private-law/json/AuthorisationCaseState.json (hunk 3 above)
Comment (khurshedk) [reply definitions/private-law/json/AuthorisationCaseState.json:364]: Changed as per provided comment

File: definitions/private-law/json/AuthorisationCaseState.json (hunk 4 above)
Comment (khurshedk) [reply definitions/private-law/json/AuthorisationCaseState.json:380]: Changed as per provided comment

File: definitions/private-law/json/CaseEvent/CaseEvent.json (hunk 2 above)
Comment (khurshedk) [reply definitions/private-law/json/CaseEvent/CaseEvent.json:1000]: checking if we can do it in backend

File: definitions/private-law/json/State.json (hunk 5 above)
Comment (khurshedk) [reply definitions/private-law/json/State.json:117]: Changed the description to 'Proceeding in offline mode in familyman system'

File: definitions/private-law/json/State.json (hunk 1 above)
Comment (khurshedk) [reply definitions/private-law/json/State.json:118]: Changed thr description to 'Proceeding in offline mode in familyman system'

**Review by khurshedk — COMMENTED** [review]
updated review comments

File: definitions/private-law/json/CaseEvent/CaseEvent.json (hunk 2 above)
Comment (AlokDatta) [reply definitions/private-law/json/CaseEvent/CaseEvent.json:1000]: anything on this?

Comment (khurshedk) [reply definitions/private-law/json/CaseEvent/CaseEvent.json]: Moved this logic to backend call

File: definitions/private-law/json/AuthorisationCaseState.json [hunk 6]
```diff
@@ -353,5 +353,36 @@
...
+          "caseworker-ras-validation",
+          "caseworker-privatelaw-solicitor",
+          "citizen",
+          "GS_profile",
+          "caseworker-privatelaw-courtadmin",
+          "caseworker-privatelaw-la",
+          "caseworker-privatelaw-judge",
+          "caseworker-caa",
+          "caseworker-approver",
+          "ctsc-team-leader",
```
[the remark this answers is not in the document]
Comment (khurshedk) [reply definitions/private-law/json/AuthorisationCaseState.json]: removed mentioned roles

**Comment by khurshedk** [comment]
Fixed Review comments

File: definitions/private-law/json/AuthorisationCaseEvent/TransferCourt/AuthorisationCaseEvent.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
+[
+  {
+    "LiveFrom": "19/01/2022",
+    "CaseTypeID": "PRLAPPS",
+    "CaseEventID": "transferToAnotherCourt",
+    "AccessControl": [
+      {
+        "UserRoles": [
+          "caseworker-privatelaw-courtadmin"
```
Comment (AlokDatta) [inline definitions/private-law/json/AuthorisationCaseEvent/TransferCourt/AuthorisationCaseEvent.json:9]: add readonly user access to event, state and case fields

File: definitions/private-law/json/CaseEvent/CaseEvent.json [hunk 8]
```diff
@@ -1087,6 +1087,25 @@
...
+    "CaseTypeID": "PRLAPPS",
+    "ID": "transferToAnotherCourt",
+    "Name": "Transfer to another court",
+    "Description": "Transfer to another court",
+    "DisplayOrder": 107,
+    "PreConditionState(s)": "CASE_ISSUED;JUDICIAL_REVIEW",
+    "PostConditionState": "*",
+    "SecurityClassification": "Public",
+    "ShowSummary": "Y",
+    "CallBackURLAboutToStartEvent": "${CCD_DEF_COS_URL}/amend-court-details/about-to-start",
```
Comment (AlokDatta) [inline definitions/private-law/json/CaseEvent/CaseEvent.json:1104]: bring the APIs in sync. make it /transfer-court.

File: definitions/private-law/json/CaseEvent/CaseEvent.json [hunk 9]
```diff
@@ -1087,6 +1087,25 @@
     "CallBackURLSubmittedEvent": "${CCD_DEF_COS_URL}/update-application",
     "ShowEventNotes": "N",
     "EndButtonLabel": "Save and continue",
+    "Publish": "True",
+    "EventEnablingCondition": "caseTypeOfApplication=\"DO_NOT_SHOW\""
+  },
+  {
+    "LiveFrom": "03/02/2022",
+    "CaseTypeID": "PRLAPPS",
+    "ID": "transferToAnotherCourt",
```
Comment (AlokDatta) [inline definitions/private-law/json/CaseEvent/CaseEvent.json:1096]: Disable amend court functionality

**Review by AlokDatta — APPROVED** [review]

**Pull request merged**